support for Twitch's new site

night opened the pull request:
I scrapped together a release for supporting the new site since you're busy.. perhaps you can improve it a tad too

night committed:
support for Twitch's new site

cletusc commented:
Thanks for your efforts getting this going. I should hopefully have time to review this pull request this holiday weekend. If it works for you then by all means update the BTTV copy to appease the masses until I've had a chance to review this.

Martii commented:
@night 

I'd start a brief review but the affected code isn't loaded in the diff because it's "too big" according to GH.

Anywho...

This line for the SPDX code:

``` js
// @license CC BY-NC-SA 3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
```

needs to be:

``` js
// @license CC-BY-NC-SA-3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
```

in order to be considered for acceptance by the server on OUJS. Same with any other "uglified" version if applicable. See https://spdx.org/licenses/CC-BY-NC-SA-3.0.html

Plus they seem to be backwards for the dual licensing... e.g.
``` js
// @license MIT; http://opensource.org/licenses/MIT
// @license CC BY-NC-SA 3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
```
... should be ...

``` js
// @license CC-BY-NC-SA-3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
// @license MIT; http://opensource.org/licenses/MIT
```

GM's metadata policy has always been to read the last key in as primary.

https://github.com/night/Userscript--Twitch-Chat-Emotes/blob/1348602ea9d8209a44c787b7d0ec2444e5332e45/package.json#L21 ... should also, at least, be SPDX compliant for some *npm* standards.

If you would be so kind to fix this in here it would be highly appreciated... pending @cletusc approval of course.

Thanks for the look. :)

night commented:
licensing issues are out of scope here.. this is not my project

Martii commented:
No problem. At least you both know about it now. :)

cletusc commented:
@Martii it won't be part of this pull request. Discussion for that should be over in #104 like you originally opened (but I hadn't replied yet).

This pull request will remain as @night originally opened (pull request should be pointed to fix one issue, e.g. not loading on site).

Martii commented:
@cletusc 
I've created so many pull requests in the last few weeks... apologies that it was duplicated.

cletusc commented on src/modules/emotes.js:
```diff
@@ -146,83 +146,49 @@ function EmoteStore() {
+					set = +set;
+					emotes.forEach(function (emote) {
+						// Set some required info.
+						emote.url = '//static-cdn.jtvnw.net/emoticons/v1/' + emote.id + '/2.0';
```
Why was this changed from `.../1.0` to `.../2.0`? The one used in chat is `.../1.0`. I see that `.../2.0` is higher quality but some of the emotes end up stretched (old emotes). Wouldn't it be better to match the quality shown in chat?

cletusc replied:
I see what Twitch is actually doing (and it's new to me). They are actually defining a [`srcset`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-srcset).

In addition to `src` being the `.../1.0`, they are also defining `srcset` as `.../1.0 1.0x, .../2.0 2.0x, .../3.0 3.0x` which picks the best image to serve based on resolution/zoom.

I'll definitely add formal support for this to get the best quality possible.

cletusc commented on src/modules/emotes.js:
```diff
@@ -333,176 +293,9 @@ function Emote(details) {
-	 * Gets the emote channel's badge image URL.
-	 * @return {string|null} The URL of the badge image for the emote's channel or `null` if it doesn't have a channel.
-	 */
-	this.getChannelBadge = function () {
```
In place of this, what do you think about using the profile image as opposed to badges considering badges aren't really seen except when on the channel's own chat whereas the profile image is used all over the place, e.g. some would be:

Mine: ![](https://static-cdn.jtvnw.net/jtv_user_pictures/cletusc-profile_image-e0c7fb2ad09fd5b0-70x70.png), DansGaming: ![](https://static-cdn.jtvnw.net/jtv_user_pictures/dansgaming-profile_image-76e4a4ab9388bc9c-70x70.png)... or others: ![](https://i.imgur.com/RknXEaq.png)

cletusc replied:
To add, this would be dead simple to add--it's already part of the data on the user API for display name and ID lookups.

night replied:
You’d have to make sure it’s in the existing API calls.. it’s expensive for us to make as many api calls as you were making

cletusc replied:
Although I agree that there were a bunch of individual calls, the badge images (along with display names) were cached in localStorage so after the first hit, it didn't hit for quite some time. With these changes, you removed both badges and display names. Badges (or alternatively like I mentioned above, the profile image), along with display names, should still be part of the menu.

The v5 API (and even the new API) allows for multiple usernames lookups in a single call, meaning display names and profile images would incur a single API call. Once you cache that, the API doesn't even get hit unless there's a new name or the user clears localStorage.

night replied:
It’s not just the act of calling an API, localstorage is slow to read from, and you individually store each. Localstorage reads and writes to disk.. I didn’t feel like refactoring it so I removed it. In the grand scheme it doesn’t matter if we have it or not

cletusc replied:
The storage API I made doesn't actually read from localStorage every time either, [it actually caches into memory](https://github.com/cletusc/storage-wrapper/blob/master/index.js#L132-L135) meaning it only reads/writes once per page load per key. This could probably be further reduced to just a few keys instead (as the storage library already supports object storage out of the box without messing with JSON).

This will definitely go back in one way or another, but I'll review how best to reduce API calls and subsequent localStorage calls.

cletusc commented on src/images/smiley.svg:
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?><svg xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:cc="http://creativecommons.org/ns#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 47.5 47.5" style="enable-background:new 0 0 47.5 47.5;" xml:space="preserve" version="1.1" id="svg2"><metadata  ...
```
Sourcing for this seems to be from [iconscout](https://iconscout.com/icon/grinning-face-smile-emoji-happy) although it's probably a common emoji image. If you have the actual source, please list it so we can add to ATTRIBUTION.md

night replied:
This is from Twitter’s open source emoji lib, https://github.com/twitter/twemoji

cletusc replied:
Seems to be a different version than [what is shown on the latest version](https://github.com/twitter/twemoji/blob/gh-pages/svg/1f603.svg): ![](https://i.imgur.com/Xyz9KLX.png)

night replied:
You are probably looking at the wrong emoji. It is from their master of when I added the asset. It should be the :D emoji

cletusc replied:
Confirmed as [1F600](https://github.com/twitter/twemoji/blob/b92cb3acb768017ba691ed9f73c3dcc49678d3d1/svg/1f600.svg):
![](https://i.imgur.com/9aLYq91.png)

cletusc commented on src/styles/style.css:
```diff
@@ -5,22 +5,34 @@
 	cursor: pointer;
 	height: 30px;
 	width: 30px;
+	-webkit-filter: grayscale(100%);
```
Just for reference: any prefixed properties (`-webkit-*`) are not needed as [autoprefixer](https://github.com/cletusc/Userscript--Twitch-Chat-Emotes/blob/cb1f75594480e65a675b9cd9186607ccb0ca71ff/gulp/tasks/styles.js#L4) is used.

Aaron128l committed:
fixes chat container selector (#2)

Aaron128l committed:
class changes (#3)

night committed:
update twitch emotes after loading them

bspotswood committed:
Ensure menu UI created on initial load (#4)

* Ensure menu UI created on initial load

This addresses a case where plugins try to load elements before the UI
dom has been created by ensuring the DOM is created up front and
attached/reattached later as needed.
https://github.com/night/BetterTTV/issues/2901

* Ensure updateEmotes() on attach()

The previous changes to replace usage of init() with new attach() method
did not include the call to updateEmotes() within attach(). This call
was made in init() and not including it can cause an issue where the
Favorites group does not show up. This change fixes that potential
issue.

* Show favorites group

Previous commit, while useful, didn't actually resolve it. Testing this
thing is so troublesome, I guess maybe I was seeing the live version
thinking it was the test. Because the UIMenu dom isn't attached when the
UIFavoritesGroup() constructor is called it can't find the
#starred-emotes-group. This changes it to look for that under the
UIMenu's dom, which will find it even while unattached.

* Fix to make resize handle work with changes

The resize handle wasn't working with the changes to how loading vs
attaching is working. Jquery UI was searching within the document for
the handle. This change passes the resize handle using a direct
reference to finding it within the UI menu dom.

night committed:
adds lazyload to emote menu

night committed:
lazy load tweaks

night committed:
remove console log

night committed:
fixes occasional error on firefox

night committed:
bugfix for react 16

night committed:
load images already in view

night committed:
make position absolute after draggable binds

night committed:
fixes duplicate button in attach

night committed:
fixes containment element selection

night committed:
attempts fixing button when viewer list goes missing

night committed:
reorder focus after value set

night committed:
move emote button

night committed:
fix tooltip

night committed:
bump dependencies

dependabot-preview[bot] committed:
[Security] Bump stringstream from 0.0.5 to 0.0.6 (#6)

night committed:
fixes button on mod view and watch parties

dependabot-preview[bot] committed:
Bump jquery from 3.5.0 to 3.5.1 (#7)

Bumps [jquery](https://github.com/jquery/jquery) from 3.5.0 to 3.5.1.
- [Release notes](https://github.com/jquery/jquery/releases)
- [Commits](https://github.com/jquery/jquery/compare/3.5.0...3.5.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

night committed:
update emotes api to graphql

night committed:
handle non-regex tokens

dependabot-preview[bot] committed:
Bump browserify from 16.5.1 to 16.5.2 (#9)

Bumps [browserify](https://github.com/browserify/browserify) from 16.5.1 to 16.5.2.
- [Release notes](https://github.com/browserify/browserify/releases)
- [Changelog](https://github.com/browserify/browserify/blob/master/changelog.markdown)
- [Commits](https://github.com/browserify/browserify/compare/v16.5.1...v16.5.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
[Security] Bump elliptic from 6.4.0 to 6.5.3 (#8)

Bumps [elliptic](https://github.com/indutny/elliptic) from 6.4.0 to 6.5.3. **This update includes a security fix.**
- [Release notes](https://github.com/indutny/elliptic/releases)
- [Commits](https://github.com/indutny/elliptic/compare/v6.4.0...v6.5.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Create Dependabot config file (#10)

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump gulp-css-base64 from 1.3.4 to 2.0.0 (#11)

Bumps [gulp-css-base64](https://github.com/zckrs/gulp-css-base64) from 1.3.4 to 2.0.0.
- [Release notes](https://github.com/zckrs/gulp-css-base64/releases)
- [Commits](https://github.com/zckrs/gulp-css-base64/compare/1.3.4...v2.0.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

night committed:
update chat container selector

night committed:
update build

dependabot[bot] committed:
Bump browserify from 16.5.2 to 17.0.0 (#12)

Bumps [browserify](https://github.com/browserify/browserify) from 16.5.2 to 17.0.0.
- [Release notes](https://github.com/browserify/browserify/releases)
- [Changelog](https://github.com/browserify/browserify/blob/master/changelog.markdown)
- [Commits](https://github.com/browserify/browserify/compare/v16.5.2...v17.0.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump watchify from 3.11.1 to 4.0.0 (#14)

Bumps [watchify](https://github.com/browserify/watchify) from 3.11.1 to 4.0.0.
- [Release notes](https://github.com/browserify/watchify/releases)
- [Changelog](https://github.com/browserify/watchify/blob/master/CHANGELOG.md)
- [Commits](https://github.com/browserify/watchify/compare/v3.11.1...v4.0.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump jquery from 3.5.1 to 3.6.0 (#15)

Bumps [jquery](https://github.com/jquery/jquery) from 3.5.1 to 3.6.0.
- [Release notes](https://github.com/jquery/jquery/releases)
- [Commits](https://github.com/jquery/jquery/compare/3.5.1...3.6.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

night committed:
fixes containment in popouts

night committed:
emote menu search (#13)

night committed:
update search styling a bit

night committed:
fixes error when emotes fail to load

dependabot[bot] committed:
Bump gulp-notify from 3.2.0 to 4.0.0 (#16)

Bumps [gulp-notify](https://github.com/mikaelbr/gulp-notify) from 3.2.0 to 4.0.0.
- [Release notes](https://github.com/mikaelbr/gulp-notify/releases)
- [Changelog](https://github.com/mikaelbr/gulp-notify/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mikaelbr/gulp-notify/commits/v4.0.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
